drivers: sensor: adxl345: implement proper interrupt configuration

kaessert opened the pull request:
Fix interrupt enable/disable sequence per ADXL345 datasheet recommendations.

The ADXL345 datasheet (page 19) states: "When initially configuring the interrupt pins, it is recommended that the functions and interrupt mapping be done before enabling the interrupts."

Main changes:

1. Remove adxl345_interrupt_config() which incorrectly enabled interrupts during driver initialization without proper sequencing. It's not clear to me what was the purpose of this function as all the required logic (setting gpio interrups, adxl mapping and enabling happens already in tigger_set). I checked other drivers and this seems to be the common pattern.

2. Implement proper INT_ENABLE disable/enable sequence in adxl345_submit_stream() when reconfiguring watermark interrupts:
   - Disable watermark interrupt in INT_ENABLE
   - Configure interrupt mapping in INT_MAP
   - Flush FIFO
   - Re-enable watermark interrupt in INT_ENABLE

3. Add datasheet-based comments in adxl345_trigger.c explaining non-obvious register operations:
   - AC/DC coupling modes (ACT_INACT_CTL register)
   - INT_MAP routing logic (0=INT1, 1=INT2)
   - Interrupt clearing by reading data registers

4. Fix bug in adxl345_trigger_set() where ADXL345_INT_MAP_DATA_RDY_MSK was hardcoded instead of using the correct int_mask variable.

5. Enhance accel_trig sample with activity detection support and improved error reporting for testing interrupt functionality.

Fixes #100891

Tested on ebyte_e73_tbb with samples `accel_polling` and `accel_trig`

stemschmidt commented on drivers/sensor/adi/adxl345/adxl345_trigger.c:
```diff
@@ -135,7 +139,7 @@ int adxl345_trigger_set(const struct device *dev,
 
 	if (handler) {
-		int_en = ADXL345_INT_MAP_DATA_RDY_MSK;
+		int_en = int_mask;
```
This is the modification that prevents the accel_trig sample to run with my given config. Your implementation is correct, but it revealed some issues which were covered by setting a hard coded ADXL345_INT_MAP_DATA_RDY_MSK.

With my configuration a 0x20 is written to FIFO_CTL. This register has a field "Samples" which is set to "0" when 0x20 is written to FIFO_CTL. The description of "Samples" has this remark: "Entering a value of 0 in the samples bits immediately
sets the watermark status bit in the INT_SOURCE register, regardless of which FIFO mode is selected."
The one bit that is set with 0x20 is the "Trigger Bit", which links the trigger event to either INT1 ("0") or INT2 ("1").
0x20 -> set watermark status bit in INT_SOURCE register, link trigger event to INT2

You modification does not set only ADXL345_INT_MAP_DATA_RDY_MSK, but also DXL345_INT_MAP_OVERRUN_MSK  and ADXL345_INT_MAP_WATERMARK_MSK (line 117-119 in this file). This int_mask is then used to map those events to INT1 (line 152) and enable the corresponding interrupts (line 173).
 
When I start the application I see that the interrupt line is immedieately asserted, but as the watermark status bit cannot be cleared (see the description above) the interrupt line remains asserted and no further interrupt is detected as it is an edge interrupt.

I have no good advice on how to solve this. I would remove DXL345_INT_MAP_OVERRUN_MSK and ADXL345_INT_MAP_WATERMARK_MSK from being set to int_mask, but double-check with Luis who had added a comment just above these lines.

ttmut commented:
Could you please fix the conflicts? Thanks.

ubieda commented on samples/sensor/accel_trig/src/main.c:
May we put the changes to the sample in another patch?

kaessert replied:
i disagree because they serve as a test too. If you want i can make a separate PR but i personally feel its very much related

ubieda reviewed (commented):
I'd like some clarity on the issue of interrupts not being cleared. Other than that and the sample-changes it looks good.

kaessert committed:
drivers: sensor: adxl345: implement proper interrupt configuration

Fix interrupt enable/disable sequence per ADXL345 datasheet
recommendations.

The ADXL345 datasheet (page 19) states: "When initially configuring
the interrupt pins, it is recommended that the functions and interrupt
mapping be done before enabling the interrupts."

Main changes:

1. Remove adxl345_interrupt_config() which incorrectly enabled
   interrupts during driver initialization without proper sequencing.
   It's not clear to me what was the purpose of this function as all
   the required logic (setting gpio interrups, adxl mapping and enabling
   happens already in tigger_set).
   I checked other drivers and this seems to be the common pattern.

2. Implement proper INT_ENABLE disable/enable sequence in
   adxl345_submit_stream() when reconfiguring watermark interrupts:
   - Disable watermark interrupt in INT_ENABLE
   - Configure interrupt mapping in INT_MAP
   - Flush FIFO
   - Re-enable watermark interrupt in INT_ENABLE

3. Add datasheet-based comments in adxl345_trigger.c explaining
   non-obvious register operations:
   - AC/DC coupling modes (ACT_INACT_CTL register)
   - INT_MAP routing logic (0=INT1, 1=INT2)
   - Interrupt clearing by reading data registers

4. Fix bug in adxl345_trigger_set() where ADXL345_INT_MAP_DATA_RDY_MSK
   was hardcoded instead of using the correct int_mask variable.

Fixes #100891

Tested on ebyte_e73_tbb with samples `accel_polling` and `accel_trig`

Signed-off-by: Tobias Kässer <t.kaesser@gmail.com>

kaessert committed:
samples: sensor: accel_trig: add activity detection and error reporting

Enhance accel_trig sample with activity detection support and
improved error reporting for testing interrupt functionality.

Signed-off-by: Tobias Kässer <t.kaesser@gmail.com>
